File: src/python-scripts/q_param.py
""" This file is able to extract q params from the q_params.txt"""
from fractions import Fraction
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple, Union

import numpy

import constants

azimuthal_2_spin_dict: Dict[str, Union[str, List[float]]] = {
    "s": "",
    "p": [0.5, 1.5],
    "d": [1.5, 2.5],
    "f": [2.5, 3.5],
}


def get_orbital(azimuthal_qn: int) -> Optional[str]:
    return list(azimuthal_2_spin_dict.keys())[azimuthal_qn]


def get_spin_as_string(azimuthal_qn: int, spin_qn: float) -> Optional[str]:
    orbital = get_orbital(azimuthal_qn)
    spins = azimuthal_2_spin_dict[orbital]

    if isinstance(spins, str):
        return spins
    elif spin_qn in spins:
        return Fraction(spin_qn).__str__()

    raise Exception(f"the {orbital} orbital cannot have a js of: {spin_qn}")


def q_param(
    data_dir: Path,
    Z: int,
    principal_qn: int,
    azimuthal_qn: int,
    spin_qn: float,
    plotter: Optional[int] = 1,
) -> Optional[Tuple[List]]:
    """Gets q_params from the q_params .txt"""

    with open(data_dir / Path("q_param.txt"), "r") as param_file:
        if plotter:
            print(
                f"{constants.Z[Z]} {principal_qn}{get_orbital(azimuthal_qn)}{get_spin_as_string(azimuthal_qn,spin_qn)}"
            )

        name = f"{Z} {principal_qn}{get_orbital(azimuthal_qn)}{get_spin_as_string(azimuthal_qn,spin_qn)}"
        print(name)
        for line in param_file:
            if line.strip() == name:
                print(line.strip())
                Eb = param_file.readline()
                Erow = numpy.fromstring(param_file.readline(), int, sep=" ")
                sigma = [Erow, numpy.fromstring(param_file.readline(), float, sep=" ")]
                beta = [Erow, numpy.fromstring(param_file.readline(), float, sep=" ")]
                gamma = [Erow, numpy.fromstring(param_file.readline(), float, sep=" ")]
                delta = [Erow, numpy.fromstring(param_file.readline(), float, sep=" ")]
                print(f"beta:{beta} gamma:{gamma} delta:{delta} Eb:{Eb}")
                return (beta, gamma, delta, Eb)
        raise NotImplementedError(f"{name} not found in params.txt")


if __name__ == "__main__":
    q_param(Path("/scratch/nal89286/xsw-script/src/fpfpp/"), 7, 2, 1, 1.5)

File: src/xsw-scripts/matlab-scripts/q_param.m
function [beta,gamma,delta,Eb] = q_param(datadir,Z,n,l,js,varargin)

if ~isempty(varargin)
    plotter = varargin{1};
else 
    plotter = 1;
end
% q_param(datadir,Z,n,l,js)
cd_old = cd;
cd(datadir)
fid = fopen('q_param.txt');
if isstring(l)~=1

    switch l
        case 0
            l='s';
        case 1

            l='p';
        case 2
            l='d';
        case 3
            l='f';
    end
   
end

switch l
    case 's'
        js = '';
    case 'p'
    	if js == 0.5
            js = '1/2';
        elseif js == 1.5
            js = '3/2';
        else
            error(['the p orbital cannot have a js of:' num2str(js)])
        end
    case 'd'
    	if js == 1.5
            js = '3/2';
        elseif js == 2.5
            js = '5/2';
        else
            error(['the d orbital cannot have a js of:' num2str(js)])            
        end
    case 'f'
    	if js == 2.5
            js = '5/2';
        elseif js == 3.5
            js = '7/2';
        else
            error(['the d orbital cannot have a js of:' num2str(js)])              
        end        
end
z={'H','He','Li','Be','B','C','N','O','F','Ne','Na','Mg','Al','Si','P','S','Cl','Ar','K','Ca','Sc','Ti','V','Cr','Mn','Fe','Co','Ni','Cu','Zn','Ga','Ge','As','Se','Br','Kr','Rb','Sr','Y','Zr','Nb','Mo','Tc','Ru','Rh','Pd','Ag','Cd','In','Sn','Sb','Te','I','Xe','Cs','Ba','La','Ce','Pr','Nd','Pm','Sm','Eu','Gd','Tb','Dy','Ho','Er','Tm','Yb','Lu','Hf','Ta','W','Re','Os','Ir','Pt','Au','Hg','Tl','Pb','Bi','Po','At','Rn','Fr','Ra','Ac','Th','Pa','U','Np','Pu','Am','Cm','Bk','Cf','Es','Fm','Md','No'};

if plotter > 0
    disp(['##############' char(10) 'Measured orbital: ' z{Z} ' ' num2str(n) l js])
end 
name = [num2str(Z) ' ' num2str(n) l js];

line = [name(1:end-1) 'Z'];
nn = 0;
%line ~= name
while strcmp(line,name)~=1
    nn = nn +1;
    line = fgetl(fid);
    while length(line) < length(name)
        line = fgetl(fid);
        nn = nn+1;
    end
end
Eb = str2num(fgetl(fid));
Erow = str2num(fgetl(fid));

sigma = [Erow;str2num(fgetl(fid))];
beta = [Erow;str2num(fgetl(fid))];
gamma = [Erow;str2num(fgetl(fid))];
delta = [Erow;str2num(fgetl(fid))];

cd(cd_old)
